diff --git a/cmake/Modules/FindCapnp_EP.cmake b/cmake/Modules/FindCapnp_EP.cmake
index d177cb0f3..08cf46fdd 100644
--- a/cmake/Modules/FindCapnp_EP.cmake
+++ b/cmake/Modules/FindCapnp_EP.cmake
@@ -56,7 +56,7 @@ if (NOT TILEDB_FORCE_ALL_DEPS OR TILEDB_CAPNP_EP_BUILT)
   endif()
 
   find_package(CapnProto
-    ${TILEDB_CAPNPROTO_VERSION} EXACT
+    ${TILEDB_CAPNPROTO_VERSION} REQUIRED
     PATHS ${TILEDB_EP_INSTALL_PREFIX}
     ${TILEDB_CAPNP_NO_DEFAULT_PATH}
     )
diff --git a/cmake/Modules/FindMagic_EP.cmake b/cmake/Modules/FindMagic_EP.cmake
index efc039d39..98724a686 100644
--- a/cmake/Modules/FindMagic_EP.cmake
+++ b/cmake/Modules/FindMagic_EP.cmake
@@ -42,6 +42,7 @@ if(TILEDB_VCPKG)
     find_path(libmagic_INCLUDE_DIR NAMES magic.h)
     find_library(libmagic_LIBRARIES magic)
     find_file(libmagic_DICTIONARY magic.mgc
+      PATHS /usr/share/file/misc
       PATH_SUFFIXES share/libmagic/misc share/misc
     )
 
@@ -107,7 +108,7 @@ if (TILEDB_LIBMAGIC_EP_BUILT)
 endif()
 
 # if not yet built add it as an external project
-if(NOT TILEDB_LIBMAGIC_EP_BUILT)
+if(false)
   if (TILEDB_SUPERBUILD)
     message(STATUS "Adding Magic as an external project")
 
diff --git a/tiledb/CMakeLists.txt b/tiledb/CMakeLists.txt
index 01db9d78d..edcf7227d 100644
--- a/tiledb/CMakeLists.txt
+++ b/tiledb/CMakeLists.txt
@@ -440,7 +440,7 @@ set(MGC_GZIPPED_BIN_INPUT_FILE "${MGC_GZIPPED_BIN_INPUT_DIRECTORY}/magic_mgc_gzi
 
 add_custom_command(
   OUTPUT "${MGC_GZIPPED_BIN_OUTPUT_FILE}"
-  DEPENDS "${MGC_GZIPPED_BIN_INPUT_FILE}"
+  DEPENDS "${MGC_GZIPPED_BIN_INPUT_FILE}" update-embedded-magic-data
   COMMAND ${CMAKE_COMMAND} -E tar x "${MGC_GZIPPED_BIN_INPUT_FILE}"
   WORKING_DIRECTORY "${MGC_GZIPPED_BIN_OUTPUT_DIRECTORY}"
 )
@@ -451,7 +451,7 @@ add_dependencies(TILEDB_CORE_OBJECTS gen_mgc_unarch)
 
 add_custom_target(
   update-embedded-magic-data
-  COMMAND "$<TARGET_FILE:tdb_gzip_embedded_data>" < "${libmagic_DICTIONARY}" "${MGC_GZIPPED_BIN_OUTPUT_FILE}"
+  COMMAND "$<TARGET_FILE:tdb_gzip_embedded_data>" < /usr/share/file/misc/magic.mgc "${MGC_GZIPPED_BIN_OUTPUT_FILE}"
   # need to work in 'local' directory with no prefix paths so no paths are included in archive
   WORKING_DIRECTORY "${MGC_GZIPPED_BIN_OUTPUT_DIRECTORY}"
   COMMAND ${CMAKE_COMMAND} -E tar cvj "magic_mgc_gzipped.bin.tar.bz2" "magic_mgc_gzipped.bin"
@@ -605,7 +605,7 @@ find_package(Magic_EP REQUIRED)
 target_link_libraries(TILEDB_CORE_OBJECTS_ILIB
   INTERFACE
     BZip2::BZip2
-    lz4::lz4
+    LZ4::lz4_shared
     spdlog::spdlog
     ZLIB::ZLIB
     ${ZSTD_TARGET}
@@ -806,6 +806,7 @@ if(TILEDB_SERIALIZATION)
     COMMAND_EXPAND_LISTS
     COMMENT "Re-generate tiledb-rest.capnp.<h,c++> files for serialization support"
     )
+  add_dependencies(TILEDB_CORE_OBJECTS update-serialization)
 
 endif()
 
diff --git a/tiledb/sm/compressors/CMakeLists.txt b/tiledb/sm/compressors/CMakeLists.txt
index b2c694c3e..7dbf44074 100644
--- a/tiledb/sm/compressors/CMakeLists.txt
+++ b/tiledb/sm/compressors/CMakeLists.txt
@@ -40,7 +40,7 @@ commence(object_library compressors)
     find_package(LZ4_EP REQUIRED)
     find_package(Zlib_EP REQUIRED)
     find_package(Zstd_EP REQUIRED)
-    this_target_link_libraries(BZip2::BZip2 lz4::lz4 ZLIB::ZLIB ${ZSTD_TARGET})
+    this_target_link_libraries(BZip2::BZip2 LZ4::lz4_shared ZLIB::ZLIB ${ZSTD_TARGET})
 conclude(object_library)
 
 #
diff --git a/tiledb/sm/filesystem/posix.cc b/tiledb/sm/filesystem/posix.cc
index 58b4f0ab1..d576cc6e0 100644
--- a/tiledb/sm/filesystem/posix.cc
+++ b/tiledb/sm/filesystem/posix.cc
@@ -47,6 +47,7 @@
 #include <sys/stat.h>
 #include <unistd.h>
 
+#include <algorithm>
 #include <fstream>
 #include <future>
 #include <iostream>
