--- a/sequence.h	2023-03-04 23:45:31.164823741 +0530
+++ b/sequence.h	2023-03-04 23:46:26.020450799 +0530
@@ -27,6 +27,7 @@
 #include <math.h>
 #include "bitmap.h"
 #include "training.h"
+#include "fptr.h"
 
 #define MAX_SEQ 32000000
 #define MAX_LINE 10000
@@ -44,9 +45,9 @@
   int end;
 } mask;
 
-int read_seq_training(FILE *, unsigned char *, unsigned char *, double *, int,
+int read_seq_training(fptr, unsigned char *, unsigned char *, double *, int,
                       mask *, int *);
-int next_seq_multi(FILE *, unsigned char *, unsigned char *, int *, double *,
+int next_seq_multi(fptr, unsigned char *, unsigned char *, int *, double *,
                    int, mask *, int *, char *, char *);
 void rcom_seq(unsigned char *, unsigned char *, unsigned char *, int);
 
@@ -85,4 +86,4 @@
 
 int imin(int, int);
 
-#endif
+#endif
\ No newline at end of file
