--- RSEM-1.3.3.orig/Makefile	2020-02-15 00:51:41.000000000 +0530
+++ RSEM-1.3.3.new/Makefile	2023-05-28 12:03:56.152930584 +0530
@@ -12,11 +12,11 @@
 
 # Compilation variables
 CXX = g++
-CXXFLAGS = -std=gnu++98 -Wall -I. -I$(BOOST) -I$(SAMTOOLS)/$(HTSLIB)
-CPPFLAGS =
+CXXFLAGS += -std=gnu++98 -Wall -I. -I$(BOOST)
+CPPFLAGS ?=
 
-LDFLAGS =
-LDLIBS =
+LDFLAGS ?=
+LDLIBS ?=
 
 # Installation variables
 INSTALL = install
@@ -30,8 +30,8 @@
 bindir = $(exec_prefix)/bin
 
 # Auxiliary variables for compilation
-SAMHEADERS = $(SAMTOOLS)/$(HTSLIB)/htslib/sam.h
-SAMLIBS = $(SAMTOOLS)/$(HTSLIB)/libhts.a
+SAMHEADERS = 
+SAMLIBS = 
 CONFIGURE = ./configure
 
 OBJS1 = parseIt.o
@@ -54,12 +54,12 @@
 
 .PHONY : all ebseq pRSEM clean
 
-all : $(PROGRAMS) $(SAMTOOLS)/samtools
+all : $(PROGRAMS)
 
 $(SAMTOOLS)/samtools :
 	cd $(SAMTOOLS) && $(CONFIGURE) --without-curses && $(MAKE) -f $(SAMTOOLS_MAKEFILE) samtools
 
-$(SAMLIBS) : $(SAMTOOLS)/samtools
+$(SAMLIBS) : 
 
 
 # Compile objects
@@ -78,10 +78,10 @@
 	$(CXX) $(LDFLAGS) -o $@ $^ $(LDLIBS)
 
 $(PROGS2) :
-	$(CXX) $(LDFLAGS) -pthread -o $@ $^ $(LDLIBS) -lz
+	$(CXX) $(LDFLAGS) -pthread -o $@ $^ $(LDLIBS) -lhts -lz
 
 $(PROGS3) :
-	$(CXX) $(LDFLAGS) -pthread -o $@ $^ $(LDLIBS)
+	$(CXX) $(LDFLAGS) -pthread -o $@ $^ $(LDLIBS) -lhts
 
 
 # Dependencies for executables
@@ -163,17 +163,13 @@
 
 
 # Install RSEM
-install : $(PROGRAMS) $(SCRIPTS) $(SAMTOOLS)/samtools rsem_perl_utils.pm
-	$(INSTALL_DIR) $(DESTDIR)$(bindir) $(DESTDIR)$(bindir)/$(SAMTOOLS)
-	$(foreach prog,$(PROGRAMS),$(INSTALL_PROGRAM) $(prog) $(DESTDIR)$(bindir)/$(prog) ; $(STRIP) $(DESTDIR)$(bindir)/$(prog) ;)
-	$(INSTALL_PROGRAM) $(SAMTOOLS)/samtools $(DESTDIR)$(bindir)/$(SAMTOOLS)/samtools
-	$(STRIP) $(DESTDIR)$(bindir)/$(SAMTOOLS)/samtools
+install : $(PROGRAMS) $(SCRIPTS) rsem_perl_utils.pm
+	$(foreach prog,$(PROGRAMS),$(INSTALL_PROGRAM) $(prog) $(DESTDIR)$(bindir)/$(prog) ; )
 	$(foreach script,$(SCRIPTS),$(INSTALL_PROGRAM) $(script) $(DESTDIR)$(bindir)/$(script) ;)
 	$(INSTALL_DATA) rsem_perl_utils.pm $(DESTDIR)$(bindir)/rsem_perl_utils.pm
 
 # Clean
 clean :
 	rm -f *.o *~ $(PROGRAMS)
-	cd $(SAMTOOLS) && $(MAKE) clean-all
 	cd EBSeq && $(MAKE) clean
 	cd pRSEM && $(MAKE) clean
