diff --git a/CMake/TeemConfig.cmake.in b/CMake/TeemConfig.cmake.in
index 348845b..04862d9 100644
--- a/CMake/TeemConfig.cmake.in
+++ b/CMake/TeemConfig.cmake.in
@@ -32,7 +32,7 @@
 # (this_dir == Teem_DIR)
 #
 GET_FILENAME_COMPONENT(this_dir "${CMAKE_CURRENT_LIST_FILE}" PATH)
-GET_FILENAME_COMPONENT(Teem_ROOT_DIR "${this_dir}/@Teem_CV_CONFIG_TO_ROOT@" ABSOLUTE)
+GET_FILENAME_COMPONENT(Teem_ROOT_DIR "${this_dir}/../../@Teem_CV_CONFIG_TO_ROOT@" ABSOLUTE)
 
 # CMake files required to build client applications that use Teem.
 SET(Teem_BUILD_SETTINGS_FILE "@Teem_CV_BUILD_SETTINGS_FILE@")
diff --git a/CMakeLists.txt b/CMakeLists.txt
index cba73d0..7c532f8 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -448,8 +448,8 @@ ENDIF(BUILD_TESTING)
 #-----------------------------------------------------------------------------
 # Help outside projects build Teem projects.
 INCLUDE(CMakeExportBuildSettings)
-EXPORT_LIBRARY_DEPENDENCIES(${Teem_BINARY_DIR}/TeemLibraryDepends.cmake)
-CMAKE_EXPORT_BUILD_SETTINGS(${Teem_BINARY_DIR}/TeemBuildSettings.cmake)
+#EXPORT_LIBRARY_DEPENDENCIES(${Teem_BINARY_DIR}/TeemLibraryDepends.cmake)
+#CMAKE_EXPORT_BUILD_SETTINGS(${Teem_BINARY_DIR}/TeemBuildSettings.cmake)
 
 SET(CFLAGS "${CMAKE_C_FLAGS}")
 SET(CC "${CMAKE_C_COMPILER}")
@@ -492,12 +492,12 @@ ELSE(Teem_USE_LIB_INSTALL_SUBDIR)
   SET(Teem_CV_CONFIG_TO_ROOT "..")
 ENDIF(Teem_USE_LIB_INSTALL_SUBDIR)
 
-SET(Teem_CV_LIBRARY_DEPENDS_FILE "\${Teem_ROOT_DIR}/lib${EXTRA_INSTALL_PATH}/TeemLibraryDepends.cmake")
+SET(Teem_CV_LIBRARY_DEPENDS_FILE "\${Teem_ROOT_DIR}/lib${EXTRA_INSTALL_PATH}/cmake/teem/TeemLibraryDepends.cmake")
 SET(Teem_CV_EXECUTABLE_DIRS "\${Teem_ROOT_DIR}/bin")
 SET(Teem_CV_LIBRARY_DIRS "\${Teem_ROOT_DIR}/lib${EXTRA_INSTALL_PATH}")
-SET(Teem_CV_USE_FILE "\${Teem_ROOT_DIR}/lib${EXTRA_INSTALL_PATH}/TeemUse.cmake")
+SET(Teem_CV_USE_FILE "\${Teem_ROOT_DIR}/lib${EXTRA_INSTALL_PATH}/cmake/teem/TeemUse.cmake")
 SET(Teem_CV_INCLUDE_DIRS "\${Teem_ROOT_DIR}/include")
-SET(Teem_CV_BUILD_SETTINGS_FILE "\${Teem_ROOT_DIR}/lib${EXTRA_INSTALL_PATH}/TeemBuildSettings.cmake")
+SET(Teem_CV_BUILD_SETTINGS_FILE "\${Teem_ROOT_DIR}/lib${EXTRA_INSTALL_PATH}/cmake/teem/TeemBuildSettings.cmake")
 SET(Teem_CV_BUILT_LIBRARIES teem) # The libraries built by teem.  Currently we only build the mega library.
 
 CONFIGURE_FILE("${Teem_SOURCE_DIR}/CMake/TeemConfig.cmake.in"
@@ -511,7 +511,7 @@ INSTALL(FILES ${Teem_INSTALLED_HEADER_FILES}
 INSTALL(FILES
   "${Teem_BINARY_DIR}/CMake/TeemConfig.cmake"
   "${Teem_SOURCE_DIR}/CMake/TeemUse.cmake"
-  "${Teem_BINARY_DIR}/TeemBuildSettings.cmake"
-  "${Teem_BINARY_DIR}/TeemLibraryDepends.cmake"
-  DESTINATION lib${EXTRA_INSTALL_PATH}
+#  "${Teem_BINARY_DIR}/TeemBuildSettings.cmake"
+#  "${Teem_BINARY_DIR}/TeemLibraryDepends.cmake"
+  DESTINATION lib${EXTRA_INSTALL_PATH}/cmake/teem
   )
