--- canu-2.3.orig/src/main.mk	2024-12-16 21:53:18.000000000 +0530
+++ canu-2.3/src/main.mk	2025-02-08 15:21:30.531526902 +0530
@@ -267,7 +267,6 @@
                pipelines/draw-tig.pl                    -> bin/draw-tig
 
 FILES       += pipelines/canu.defaults                  -> bin/canu.defaults \
-               mhap/mhap-2.1.3.jar                      -> share/java/classes/mhap-2.1.3.jar \
                pipelines/canu/Consensus.pm              -> lib/perl5/site_perl/canu/Consensus.pm \
                pipelines/canu/CorrectReads.pm           -> lib/perl5/site_perl/canu/CorrectReads.pm \
                pipelines/canu/HaplotypeReads.pm         -> lib/perl5/site_perl/canu/HaplotypeReads.pm \
