Author: Andreas Tille <tille@debian.org>
Last-Update: Wed, 12 Oct 2022 18:03:54 +0200
Description: C++ comments are invalid in ANSI C code

--- a/src/mcl/alg.c
+++ b/src/mcl/alg.c
@@ -1119,7 +1119,7 @@ mcxstatus mclAlgorithmInit
 
             case ALG_OPT_AMOIXA
          :  helpbits |= MCX_OPT_DISPLAY_HIDDEN  ;
-            // fall through
+            /* fall through */
             case ALG_OPT_SHOWLONGHELP
          :  helpbits |= MCX_OPT_DISPLAY_SKIP
 
--- a/src/shmx/mcxload.c
+++ b/src/shmx/mcxload.c
@@ -706,9 +706,9 @@ int main
          ;
 
             case MY_OPT_STRICT_TABG :   t++ ;
-            // fall through
+            /* fall through */
             case MY_OPT_RESTRICT_TABG : t++ ;
-            // fall through
+            /* fall through */
             case MY_OPT_EXTEND_TABG
          :
             {  xfusetabg = mcxIOnew(opt->val, "r")
@@ -720,9 +720,9 @@ int main
          ;  }
 
             case MY_OPT_STRICT_TABC :   t++ ;
-            // fall through
+            /* fall through */
             case MY_OPT_RESTRICT_TABC : t++ ;
-            // fall through
+            /* fall through */
             case MY_OPT_EXTEND_TABC
          :
             {  xfusetabc = mcxIOnew(opt->val, "r")
@@ -734,9 +734,9 @@ int main
          ;  }
 
             case MY_OPT_STRICT_TABR :   t++ ;
-            // fall through
+            /* fall through */
             case MY_OPT_RESTRICT_TABR : t++ ;
-            // fall through
+            /* fall through */
             case MY_OPT_EXTEND_TABR
          :
             {  xfusetabr = mcxIOnew(opt->val, "r")
