--- a/configure.ac
+++ b/configure.ac
@@ -6,7 +6,6 @@
 AC_CONFIG_SRCDIR([jellyfish])
 AC_CONFIG_HEADERS([config.h])
 LT_INIT
-AC_LIB_RPATH
 PKG_PROG_PKG_CONFIG
 
 # Change default compilation flags
@@ -28,8 +27,7 @@
 AC_ARG_ENABLE([htslib],
               [AS_HELP_STRING([--enable-htslib], [Look for the HTS library (default=yes)])])
 AS_IF([test "x$enable_htslib" = "xyes" -o "x$enable_htslib" = "x"],
-      [PKG_CHECK_MODULES([HTSLIB], [htslib], [AC_DEFINE([HAVE_HTSLIB], [1], [Defined if htslib is available])], [true])]
-      [AC_LIB_LINKFLAGS_FROM_LIBS([HTSLIB_RPATH], [$HTSLIB_LIBS], [LIBTOOL])])
+      [PKG_CHECK_MODULES([HTSLIB], [htslib], [AC_DEFINE([HAVE_HTSLIB], [1], [Defined if htslib is available])], [true])])
 # Look for sammtools
 AC_ARG_VAR([SAMTOOLS], [Path to samtools program])
 AS_IF([test "x$SAMTOOLS" = "x"],
