--- jellyfish-2.3.1/swig/Makefile.am	2019-07-13 02:07:53.000000000 +0530
+++ jellyfish-2.3.1.new/swig/Makefile.am	2023-12-10 17:22:15.902052082 +0530
@@ -20,11 +20,6 @@
 PYTHON_BUILT = swig/python/swig_wrap.cpp swig/python/dna_jellyfish.py
 BUILT_SOURCES += $(PYTHON_BUILT)
 
-if PYTHON_DEPRECATED
-pythonglobaldir = $(PYTHON_SITE_PKG)
-pythonglobal_SCRIPTS = swig/python/jellyfish.py
-EXTRA_DIST += $(pythonglobal_SCRIPTS)
-endif
 pythonextdir = $(PYTHON_SITE_PKG)/dna_jellyfish
 pythonext_SCRIPTS = swig/python/__init__.pyc
 pythonext_LTLIBRARIES = swig/python/_dna_jellyfish.la
