--- a/setup.py	2023-05-14 00:15:53.983680836 +0530
+++ b/setup.py	2023-05-14 00:17:57.185420794 +0530
@@ -63,12 +63,14 @@
         "ncls.src.fncls", ["ncls/src/fncls.pyx", "ncls/src/fintervaldb.c"],
         # define_macros=macros,
         include_dirs=include_dirs)]
+for e in extensions:
+    e.cython_directives = {'language_level': "3"}
 
 # using setuptools to cythonize if cython not found
 # not recommended by cython docs, but still
 try:
     from Cython.Build import cythonize
-    ext_modules = cythonize(extensions, language_level=2)
+    ext_modules = cythonize(extensions, compiler_directives={'language_level' : "3"})
 except ImportError:
     print()
     print("Warning: Cannot compile with Cython. Using legacy build.")
