--- a/source/uchime_src/makefile
+++ b/source/uchime_src/makefile
@@ -1,6 +1,7 @@
 CXXFLAGS = -O3 -D_FILE_OFFSET_BITS=64 -DNDEBUG=1 -DUCHIMES=1
 LDFLAGS = -g
-
+CXXFLAGS += -g -fstack-protector-strong -Wformat -Werror=format-security
+LDFLAGS += -Wl,-z,relro -Wl,-z,now
 #
 # Get the list of all .cpp files, rename to .o files
 #
@@ -10,7 +11,7 @@
     OBJECTS+=$(patsubst %.c,%.o,$(wildcard *.c))
 
 uchime : $(OBJECTS)
-	$(CXX) $(LDFLAGS) $(TARGET_ARCH) -o $@ $(OBJECTS) $(LIBS)
+	$(CXX) $(CXXFLAGS) $(LDFLAGS) $(TARGET_ARCH) -o $@ $(OBJECTS) $(LIBS)
 
 install : uchime
 
@@ -26,4 +27,4 @@
 
 clean :
 	@rm -f $(OBJECTS)
-	
\ No newline at end of file
+	
