From: Andreas Tille <tille@debian.org>
Date: Mon, 20 Jun 2022 13:56:45 +0200
Subject: uchime code does not fit c++17 standard

Bug-Debian: https://bugs.debian.org/984243
Last-Update: Thu, 21 Oct 2021 11:22:32 +0200
Origin: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=984243#33
---
 source/uchime_src/makefile | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

--- a/source/uchime_src/makefile
+++ b/source/uchime_src/makefile
@@ -1,4 +1,4 @@
-CXXFLAGS = -O3 -D_FILE_OFFSET_BITS=64 -DNDEBUG=1 -DUCHIMES=1
+CXXFLAGS = -O3 -D_FILE_OFFSET_BITS=64 -DNDEBUG=1 -DUCHIMES=1 -std=c++14
 LDFLAGS = -g
 CXXFLAGS += -g -fstack-protector-strong -Wformat -Werror=format-security
 LDFLAGS += -Wl,-z,relro -Wl,-z,now
