From: Andreas Tille <tille@debian.org>
Date: Mon, 20 Jun 2022 13:56:45 +0200
Subject: Fix build with gcc-11

Bug-Debian: https://bugs.debian.org/984243
Last-Update: Thu, 21 Oct 2021 11:22:32 +0200
---
 source/writer.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/source/writer.h b/source/writer.h
index 06f4f1b..cd1142f 100755
--- a/source/writer.h
+++ b/source/writer.h
@@ -10,6 +10,7 @@
 #define writer_h
 
 #include "sharedwriter.hpp"
+#include <memory>
 
 /***********************************************************************/
 
