--- minimap2-2.26.orig/Makefile	2023-04-29 21:51:09.000000000 +0530
+++ minimap2-2.26.new/Makefile	2023-10-21 16:56:41.816040016 +0530
@@ -1,5 +1,6 @@
-CFLAGS=		-g -Wall -O2 -Wc++-compat #-Wextra
-CPPFLAGS=	-DHAVE_KALLOC
+CFLAGS+=		-g -Wall -fPIC -pie -O2 -Wc++-compat #-Wextra
+CPPFLAGS+=	-DHAVE_KALLOC
+LDFLAGS += -pie
 INCLUDES=
 OBJS=		kthread.o kalloc.o misc.o bseq.o sketch.o sdust.o options.o index.o \
 			lchain.o align.o hit.o seed.o map.o format.o pe.o esterr.o splitidx.o \
@@ -40,6 +41,8 @@
 
 .PHONY:all extra clean depend
 .SUFFIXES:.c .o
+NAME := minimap2
+LIBRARY := lib$(NAME).so
 
 .c.o:
 		$(CC) -c $(CFLAGS) $(CPPFLAGS) $(INCLUDES) $< -o $@
@@ -48,17 +51,17 @@
 
 extra:all $(PROG_EXTRA)
 
-minimap2:main.o libminimap2.a
-		$(CC) $(CFLAGS) main.o -o $@ -L. -lminimap2 $(LIBS)
+$(LIBRARY): $(OBJS)
+		$(CC) $(LDFLAGS) -shared -Wl,-soname,lib$(NAME).so $^ -lm -lz -o $@
 
-minimap2-lite:example.o libminimap2.a
-		$(CC) $(CFLAGS) $< -o $@ -L. -lminimap2 $(LIBS)
+minimap2:main.o libminimap2.so
+		$(CC) $(LDFLAGS) main.o -o $@ -L. -lminimap2 $(LIBS)
 
-libminimap2.a:$(OBJS)
-		$(AR) -csru $@ $(OBJS)
+minimap2-lite:example.o libminimap2.so
+		$(CC) $(LDFLAGS) $< -o $@ -L. -lminimap2 $(LIBS)
 
 sdust:sdust.c kalloc.o kalloc.h kdq.h kvec.h kseq.h ketopt.h sdust.h
-		$(CC) -D_SDUST_MAIN $(CFLAGS) $< kalloc.o -o $@ -lz
+		$(CC) $(LDFLAGS) -D_SDUST_MAIN $(CFLAGS) $< kalloc.o -o $@ -lz
 
 # SSE-specific targets on x86/x86_64
 
